grr/server/grr_response_server/flows/general/osquery_test.py

```diff
@@ -261,11 +268,182 @@ def testTotalRowCountIncludesAllChunks(self):
+        fd.write(b"1" * target_bytes)
+
+      table = f"""
+      [
+        {{ "collect_collumn": "{temp_file_path}" }}
+      ]
+      """
+
+      with mock.patch.object(
+          osquery_flow,
+          "FILE_COLLECTION_MAX_TOTAL_BYTES",
+          less_than_necessary_bytes):
+        with osquery_test_lib.FakeOsqueryiOutput(stdout=table, stderr=""):
+          with self.assertRaises(RuntimeError):
+            self._RunFlow("Doesn't matter", ["collect_collumn"])
+
+  def testFlowCollectFile(self):
+    with temp.AutoTempFilePath() as temp_file_path:
+      with io.open(temp_file_path, "wb") as fd:
+        fd.write("Just sample text to put in the file.".encode('utf-8'))
+
+      table = f"""
+      [
+        {{ "collect_column": "{temp_file_path}" }}
+      ]
+      """
+
+      with osquery_test_lib.FakeOsqueryiOutput(stdout=table, stderr=""):
+        results = self._RunFlow("Doesn't matter", ["collect_column"])
+      
+    self.assertLen(results, 2)
+    self.assertEqual(type(results[0]), rdf_osquery.OsqueryResult)
+    self.assertEqual(type(results[1]), rdf_client_fs.StatEntry)
+
+    pathspec = results[1].pathspec
+    client_path = db.ClientPath.FromPathSpec(self.client_id, pathspec)
+    fd_rel_db = file_store.OpenFile(client_path)
+    file_text = fd_rel_db.read().decode("utf-8")
+    self.assertEqual(file_text, "Just sample text to put in the file.")
+
+  def testFlowDoesntCollectWhenColumnsAboveLimit(self):
+    with temp.AutoTempFilePath() as temp_file_path:
+      with io.open(temp_file_path, "wb") as fd:
+        fd.write("Just sample text to put in the file.".encode('utf-8'))
+
+      with mock.patch.object(osquery_flow, "FILE_COLLECTION_MAX_COLUMNS", 1):
+        # Should raise immediately, no need to fake Osquery output
+        with self.assertRaises(RuntimeError):
+          self._RunFlow("Doesn't matter", ["collect1", "collect2"])
+
+  def testFlowDoesntCollectWhenRowsAboveLimit(self):
+    with temp.AutoTempFilePath() as temp_file_path:
+      with io.open(temp_file_path, "wb") as fd:
+        fd.write("Just sample text to put in the file.".encode('utf-8'))
+
+      table = f"""
+      [
+        {{ "collect_column": "{temp_file_path}"}},
+        {{ "collect_column": "{temp_file_path}"}}
+      ]
+      """
+
+      with mock.patch.object(osquery_flow, "FILE_COLLECTION_MAX_ROWS", 1):
+        with osquery_test_lib.FakeOsqueryiOutput(stdout=table, stderr=""):
+          with self.assertRaises(RuntimeError):
+            self._RunFlow("Doesn't matter", ["collect_collumn"])
+
+  def testArchiveMappingsForMultipleFiles(self):
+    with temp.AutoTempFilePath() as temp_file_path1, \
+         temp.AutoTempFilePath() as temp_file_path2:
+      with io.open(temp_file_path1, "wb") as fd:
+        fd.write("Just sample text to put in the file 1.".encode('utf-8'))
+      with io.open(temp_file_path2, "wb") as fd:
+        fd.write("Just sample text to put in the file 2.".encode('utf-8'))
```

Comment:
You can use:

```python
with temp.AutoTempDirPath(remove_non_empty=True) as tempdir_path:
    temp_filepath_1 = os.path.join(tempdir_path, "foo")
    temp_filepath_2 = os.path.join(tempdir_path, "bar")
    (...)
```

This uses only one context manager with temporary state. And, as I have learned the hard way myself, is also easier to debug in case the test fails—you hit a breakpoint, go to the temporary folder and can observe the state of the only two files there rather than navigating through thousand garbage files in your `/tmp`.

Reply:
That's a good point, thanks.